### PR TITLE
Carcará X Image Processing IOC

### PR DESCRIPTION
**Pull request by RafaelLyra8**

### CODE REVIEW

File: cax-ap-cam/cax_ap_cam/const.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from siriuspy.envars import VACA_PREFIX as _vaca_prefix
+
+class Constants:
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/const.py:5]: I think this class can derivate from `siriuspy.csdev.Const`, and can be renamed to `Const` or `ImgProcConst`

File: cax-ap-cam/cax_ap_cam/const.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+from siriuspy.envars import VACA_PREFIX as _vaca_prefix
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/const.py:1]: This module can be renamed to `csdev`, as we generally use in other IOCs

File: cax-ap-cam/README.md [hunk 3]
```diff
@@ -1,2 +1,2 @@
-# li-ap-energy
+# cax-ap-cam
```
Comment (anacso17) [inline cax-ap-cam/README.md:1]: I suggest bl-ap-imgproc for package and IOC name

File: cax-ap-cam/INSTALL [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+This package can be installed using setuptools (>=16.0). There is no automatic
+dependency checking, so read the list below and ensure necessary dependencies
+are available. Developed and tested with Python 3 (3.6).
+
+
+REQUIREMENTS
+============
+
+- pcaspy 0.65
```
Comment (anacso17) [inline cax-ap-cam/INSTALL:12]: we use pcaspy 0.7.3

File: cax-ap-cam/INSTALL [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+
+INSTALLATION
+============
+
+To install to a default Python directory, run
+
+    <python> setup.py install
```
Comment (anacso17) [inline cax-ap-cam/INSTALL:20]: `make install`

File: cax-ap-cam/INSTALL [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+Use the appropriate command in <python>. Depending on permissions, it may be
+necessary to run the command with sudo or as superuser.
+
+To install a development version, with a link to the local package files in the
+default Python directory, run
+
+    <python> setup.py develop
```
Comment (anacso17) [inline cax-ap-cam/INSTALL:28]: `make develop-install`

File: cax-ap-cam/LICENSE [hunk 7]
```diff
@@ -0,0 +1,674 @@
+GNU GENERAL PUBLIC LICENSE
```
Comment (anacso17) [inline cax-ap-cam/LICENSE:1]: this file should be a link to the global LICENSE file

File: cax-ap-cam/build.sh [hunk 8]
```diff
@@ -0,0 +1,3 @@
+make
```
Comment (anacso17) [inline cax-ap-cam/build.sh:1]: I don't think we need this build.sh.

File: cax-ap-cam/cax_ap_cam/cax_ap_cam.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        pvname=ioc_prefix + sorted(db.keys())[0],
+        use_prefix=False, timeout=0.5)
+
+    if running:
+        _log.error('Another ' + ioc_prefix + ' is already running!')
+        return True
+
+    _util.print_ioc_banner(
+        '', db, 'Image Processing IOC.',
+        db['Version-Cte']['value'], ioc_prefix)
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/cax_ap_cam.py:103]: I think this part should be in run_imag_proc

File: cax-ap-cam/cax_ap_cam/cax_ap_cam.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    _log.info('Creating Driver.')
+    app = App(const=const)
+    _Driver(app)
+    return app
+
+
+def ioc_main_loop(app):
+    interval = 0.5
+    while not stop_event.is_set():
+        app.process(interval)
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/cax_ap_cam.py:147]: I think this part should be in run_imag_proc

File: cax-ap-cam/cax_ap_cam/const.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+                    'type': 'float', 'unit': 'intensity'
+                },
+                'ROI' + axis + '_Fit_Error' + sufix: {
+                    'type': 'float', 'unit': '%'
+                }
+            })
+
+        return db
+
+    def set_database(self):
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/const.py:67]: I think this should be a private method

File: cax-ap-cam/cax_ap_cam/const.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+                'type': 'int', 'unit': 'intensity'
+            },
+            'Intensity_Avg' + sufix: {
+                'type': 'float', 'unit': 'intensity'
+            },
+            'Intensity_Sum' + sufix: {
+                'type': 'int', 'unit': 'intensity'
+            },
+            'Is_Saturated' + sufix: {
+                'type': 'int'
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/const.py:29]: I think we can remove those "_" in PV names: `Intensity_Min` -> `IntensityMin`

File: cax-ap-cam/cax_ap_cam/image_proc.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+        data = images.Image2D.generate_gaussian_2d(
+            amp=255, angle=0*np.pi/180, offset=0, rand_amp=0,
+            saturation_intensity=None,
+            sizex=1280, sigmax=50, meanx=500,
+            sizey=1024, sigmay=20, meany=600,
+        )
+        self.imag_processor = images.Image2D_Fit(data=data)
+
+    def pvname2attrname(self, pvname):
+        return pvname[:-3].lower()
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/image_proc.py:15]: I think this mapping should be in App class

File: cax-ap-cam/cax_ap_cam/image_proc.py [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+    def update_roi(self, reason, value, update_pvs):
+        if 'X' in reason:
+            self.imag_processor.roix = value
+        else:
+            self.imag_processor.roiy = value
+        update_pvs(self.imag_processor)
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/image_proc.py:22]: I think PV manipulations should be in App class

File: cax-ap-cam/cax_ap_cam/const.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+                    'type': 'float', 'unit': 'px'
+                },
+                'ROI' + axis + '_Mean' + sufix: {
+                    'type': 'float', 'unit': 'px'
+                },
+                'ROI' + axis + '_Amplitude' + sufix: {
+                    'type': 'float', 'unit': 'intensity'
+                },
+                'ROI' + axis + '_Fit_Error' + sufix: {
+                    'type': 'float', 'unit': '%'
```
Comment (anacso17) [inline cax-ap-cam/cax_ap_cam/const.py:61]: Also, remove those "_" in PV names here

File: cax-ap-cam/cax_ap_cam/const.py (hunk 1 above)
Comment (xresende) [reply cax-ap-cam/cax_ap_cam/const.py]: @anacso17 , I have added more commits to this module, trying to standardize it. it should be moved to siriuspy soon.

File: cax-ap-cam/cax_ap_cam/const.py (hunk 2 above)
Comment (xresende) [reply cax-ap-cam/cax_ap_cam/const.py]: agreed.

File: cax-ap-cam/README.md (hunk 3 above)
Comment (xresende) [reply cax-ap-cam/README.md]: done.

File: cax-ap-cam/INSTALL (hunk 4 above)
Comment (xresende) [reply cax-ap-cam/INSTALL]: done

File: cax-ap-cam/LICENSE (hunk 7 above)
Comment (xresende) [reply cax-ap-cam/LICENSE]: done

File: cax-ap-cam/build.sh (hunk 8 above)
Comment (xresende) [reply cax-ap-cam/build.sh]: agreed. removed.

File: cax-ap-cam/cax_ap_cam/cax_ap_cam.py (hunk 9 above)
Comment (xresende) [reply cax-ap-cam/cax_ap_cam/cax_ap_cam.py]: Igreed. done.

File: cax-ap-cam/cax_ap_cam/const.py (hunk 11 above)
Comment (xresende) [reply cax-ap-cam/cax_ap_cam/const.py]: I think it should not exist. I deleted it.

File: cax-ap-cam/cax_ap_cam/const.py (hunk 12 above)
Comment (xresende) [reply cax-ap-cam/cax_ap_cam/const.py]: done

File: cax-ap-cam/cax_ap_cam/image_proc.py (hunk 13 above)
Comment (xresende) [reply cax-ap-cam/cax_ap_cam/image_proc.py]: deprecated

File: cax-ap-cam/cax_ap_cam/const.py (hunk 15 above)
Comment (xresende) [reply cax-ap-cam/cax_ap_cam/const.py]: done.

File: cax-ap-cam/cax_ap_cam/image_proc.py (hunk 14 above)
Comment (xresende) [reply cax-ap-cam/cax_ap_cam/image_proc.py]: yes. done.

File: bl-ap-imgproc/INSTALL [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+This package can be installed using setuptools (>=16.0). There is no automatic
+dependency checking, so read the list below and ensure necessary dependencies
+are available. Developed and tested with Python 3 (3.6).
+
+
+REQUIREMENTS
+============
+
+- pcaspy 0.65
```
Comment (anacso17) [inline bl-ap-imgproc/INSTALL:12]: wrong version of pcaspy: 0.7.3

File: bl-ap-imgproc/INSTALL [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+    make install
+
+Depending on permissions, it may be
+necessary to run the command with sudo or as superuser.
+
+To install a development version, with a link to the local package files in the
+default Python directory, run
+
+    make install-develop
```
Comment (anacso17) [inline bl-ap-imgproc/INSTALL:28]: error: develop-install

File: requirements.txt [hunk 18]
```diff
@@ -1,3 +1,2 @@
-pcaspy>=0.7.2
-PRUserial485>=2.5.0
+pcaspy>=0.7.3
 siriuspy>=2.2.0
```
Comment (anacso17) [inline requirements.txt:2]: maybe it is interesting to define mathphys minimum version here

File: bl-ap-imgproc/bl_ap_imgproc/meas.py [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+
+        except Exception:
+            self._update_success = False
+
+        # run registered driver callback
+        if self._callback:
+            self._callback()
+
+    def _create_dvf(self):
+        """Create DVO object and add process_image callback."""
```
Comment (anacso17) [inline bl-ap-imgproc/bl_ap_imgproc/meas.py:190]: typo? DVF

File: bl-ap-imgproc/INSTALL (hunk 16 above)
Comment (xresende) [reply bl-ap-imgproc/INSTALL]: done

File: bl-ap-imgproc/INSTALL (hunk 17 above)
Comment (xresende) [reply bl-ap-imgproc/INSTALL]: done

File: bl-ap-imgproc/bl_ap_imgproc/meas.py (hunk 19 above)
Comment (xresende) [reply bl-ap-imgproc/bl_ap_imgproc/meas.py]: done

File: requirements.txt (hunk 18 above)
Comment (xresende) [reply requirements.txt:3]: done

**Review by xresende — APPROVED** [review]

File: bl-ap-imgproc/bl_ap_imgproc/bl_ap_imgproc.py [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+            return False
+        elif self.check_value_none(value):
+            return False
+        elif self.check_enums_nok(reason, value):
+            return False
+        else:
+            return True
+
+
+def defineAbortFunction():
```
Comment (fernandohds564) [inline bl-ap-imgproc/bl_ap_imgproc/bl_ap_imgproc.py:92]: consider using `_` instead of camelCase letters.

**Review by fernandohds564 — CHANGES_REQUESTED** [review]
I think most of these codes should be in `siriuspy`, shouldn't they?

File: bl-ap-imgproc/bl_ap_imgproc/bl_ap_imgproc.py (hunk 20 above)
Comment (xresende) [reply bl-ap-imgproc/bl_ap_imgproc/bl_ap_imgproc.py:92]: yes, @fernandohds564 , @RafaelLyra8 is still learning our lib structures with this project. I think we can standardize gradually so as not to introduce a too steep learning curve to be overcome that hinders productivity. but simple suggestions I can implement myself in this PR and can instruct @RafaelLyra8.

**Comment by xresende** [comment]
> > I think most of these codes should be in `siriuspy`, shouldn't they?
> 
> I think that `const.py` (could also be renamed to `csdev.py`), `main.py` and `meas.py` could actually be in the `siriuspy.meas` subpackage

same comment above, as I gave to @fernandohds564, @anacso17 .

**Review by anacso17 — APPROVED** [review]

File: bl-ap-imgproc/bl_ap_imgproc/bl_ap_imgproc.py (hunk 20 above)
Comment (xresende) [reply bl-ap-imgproc/bl_ap_imgproc/bl_ap_imgproc.py:92]: done

**Comment by xresende** [comment]
> > I think most of these codes should be in `siriuspy`, shouldn't they?
> 
> I think that `const.py` (could also be renamed to `csdev.py`), `main.py` and `meas.py` could actually be in the `siriuspy.meas` subpackage

renames module from `const.py`to `csdev.py`

**Comment by xresende** [comment]
> > > I think most of these codes should be in `siriuspy`, shouldn't they?
> > 
> > 
> > I think that `const.py` (could also be renamed to `csdev.py`), `main.py` and `meas.py` could actually be in the `siriuspy.meas` subpackage
> 
> renames module from `const.py`to `csdev.py`

done

**Review by anacso17 — APPROVED** [review]

**Review by fernandohds564 — APPROVED** [review]

**Pull request merged**